src/lib/Parser.hs

```diff
@@ -260,6 +264,52 @@ topLet = do
         k = if v == mkName "eff" then EffectRowKind else TypeKind
         uTyKind = WithSrc (Just pos) $ UPrimExpr $ TCExpr k
 
+interfaceDef :: Parser [UDecl]
+interfaceDef = do
+  keyWord InterfaceKW
+  -- TODO: allow multi-parameter type classes
+  interfaceName <- upperName <|> symName
+  (oneAnnBinder, pos) <- withPos annBinder
+  keyWord WhereKW
+  record <- withSrc $ URecordTy <$> interfaceRecordFields ":"
+  let consName = mkInterfaceConsName interfaceName
+  let (varTypeName, oneAnnBinder') = case oneAnnBinder of
+        Bind v    -> (varName v, oneAnnBinder)
+        Ignore ty -> let genVar = mkName $ mkNoShadowingStr "gen" in
+            (genVar, Bind (genVar:>ty))
+  let tyCon = UConDef interfaceName $ toNest [oneAnnBinder']
+  let funDefs = mkFunDefs (pos, varTypeName, interfaceName) record
+  return $ (UData tyCon [UConDef consName (toNest [Ignore record])]) : funDefs
+  where
+    mkFunDefs :: (SrcPos, Name, Name) -> UExpr -> [UDecl]
+    mkFunDefs meta ~(WithSrc _ (URecordTy (NoExt (LabeledItems items)))) =
+        fmap (\(name, ty :| []) -> mkOneFunDef meta (name, ty)) $ M.toList items
+    mkOneFunDef :: (SrcPos, Name, Name) -> (Label, UExpr) -> UDecl
+    mkOneFunDef (pos, varTypeName, interfaceName) (fLabel, fType) =
+      let (p, ann) = lhs
+      in let (ann', rhs') = addImplicitImplicitArgs pos ann rhs
+      in ULet PlainLet (p, ann') rhs'
+      where
+        lhs = let uAnnBinder = Bind $ instanceName :> ns (UApp
+                    (PlainArrow ()) (var interfaceName) (var varTypeName))
+          in (patb fLabel, Just $ ns $ UPi uAnnBinder ClassArrow fType)
+        rhs =
+          let recordStr = mkNoShadowingStr "recordVar"
```

Comment:
nit: this is a local variable, right, so we don't need to worry about shadowing?